### PR TITLE
[Iwyu_tool] Remove -Xiwyu implicit prefix

### PR DESCRIPTION
**Pull request by alacasta**
# What?

iwyu_tool included `-Xiwyu` prefix to all those arguments comming after the separator `--`. While this had the advantage of not requiring users to provide this flag explicitely, the assumption was that all those arguments were `iwyu` arguments. 

Considering the current iwyu interface 

```
include-what-you-use [-Xiwyu --iwyu_opt]... <clang opts> <source file>
```

The change purposed here goes in the direction of removing this implicitely setting in favor of being able to provide also other type of arguments that iwyu support like `<clang opts>` 

Issue: iwyu_tool.py does not attend to clang flags #612

# How?

The previous implicit handling of the arguments

```
    # Force -Xiwyu prefix to iwyu_args so users don't have to provide prefix
    # explicitly.
    prefixes = ['-Xiwyu'] * len(iwyu_args)
    iwyu_args = list(sum(zip(prefixes, iwyu_args), ()))
```

has been removed so that `iwyu_args` are passed verbatim

Additional tests covering how the argumentes are processed have been also included

### CODE REVIEW

**Comment by alacasta** [comment]
CI is not able to run these tests because of #630

**Comment by alacasta** [comment]
Ci is now running fine once the Mock dependency has been removed. An alternative monkey patch implementation is proposed.

File: iwyu_tool.py [hunk 1]
```diff
@@ -435,12 +435,6 @@ def partition_args(argv):
             return argv, []
     argv, iwyu_args = partition_args(sys.argv[1:])
```
Comment (kimgr) [inline iwyu_tool.py]: It would be nice to rename `iwyu_args` to `extra_args` now that they're not strictly -Xiwyu args.

But please make that refactor a separate commit, as it touches several places in this file.

File: iwyu_tool_test.py [hunk 2]
```diff
@@ -52,6 +52,57 @@ def will_return(self, content):
...
+        self._will_return = ''
+        self._call_args = []
+
+    def reset(self):
+        iwyu_tool.main = self.real_iwyu_tool_main
+
+    def get_call_args(self):
+        return self._call_args
+
+    def will_return(self, content):
```
Comment (kimgr) [inline iwyu_tool_test.py]: This is unused, and can be removed

File: iwyu_tool_test.py [hunk 3]
```diff
@@ -52,6 +52,57 @@ def will_return(self, content):
...
+
+class MockIwyuToolMain(object):
+    """ Provides an object that configures a mock'd behaviour
+        of the iwyu_tool.main method, being possible to obtain
+        the arguments used in the called within the section tested.
+    """     
+    def __init__(self):
+        self.real_iwyu_tool_main = iwyu_tool.main
+        iwyu_tool.main = self._iwyu_tool_main_mock
+        self._will_return = ''
```
Comment (kimgr) [inline iwyu_tool_test.py]: Unused

File: iwyu_tool_test.py [hunk 4]
```diff
@@ -208,6 +259,40 @@ def test_consecutive_spaces(self):
...
+        self.iwyu_tool_main_mock =  MockIwyuToolMain()
+        self.sys_exit_stub = StubSysExit()
+        self.sys_exit_stub.will_return(0)
+
+    def tearDown(self):
+        self.iwyu_tool_main_mock.reset()
+        self.sys_exit_stub.reset()
+
+    def test_argument_parser_sets_argument_correctly(self):
+        """ Check that arguments are verbatim injected to iwyu. """
```
Comment (kimgr) [inline iwyu_tool_test.py]: verbatim injected -> injected verbatim

File: iwyu_tool_test.py [hunk 5]
```diff
@@ -208,6 +259,40 @@ def test_consecutive_spaces(self):
...
+        argv_stub = StubSysArgv(["iwyu_tool.py", "-p", ".", "--", "arg1"])
+        iwyu_tool._bootstrap()
+        iwyu_call_args = self.iwyu_tool_main_mock.get_call_args()
+        self.assertIn(["arg1"], iwyu_call_args)
+
+    def test_argument_parser_does_include_xiwyu_flag(self):
+        """ Check that -Xiwyu and --iwyu_opt arguments are passed to the iwyu call. """
```
Comment (kimgr) [inline iwyu_tool_test.py]: `--iwyu_opt` has no special meaning for IWYU, so it probably doesn't need to be called out in the test here. The important quality is that `-Xiwyu --something` is forwarded.

File: iwyu_tool_test.py [hunk 6]
```diff
@@ -52,6 +52,57 @@ def will_return(self, content):
...
+        iwyu_tool.main = self.real_iwyu_tool_main
+
+    def get_call_args(self):
+        return self._call_args
+
+    def will_return(self, content):
+        self._will_return = content      
+
+    def _iwyu_tool_main_mock(self, *args):
+        self._call_args = args
```
Comment (kimgr) [inline iwyu_tool_test.py]: This would be stronger if you could remember *which* arg got which value.

I can't see an elegant way to do that, so the easiest is probably to repeat `main`'s argument list here, and store the args in individual named fields;

```
def main(self, compilation_db_path, source_files, verbose, formatter, jobs, iwyu_args):
    self.compilation_db_path = compilation_db_path
    self.source_files = source_files
    self.verbose = verbose
    self.formatter = formatter
    self.jobs = jobs
    self.iwyu_args = iwyu_args
    return 0
```

That way the tests become much more specific, e.g.

    self.assertEqual(['-Xiwyu', '--arg'], mock.iwyu_args);

File: iwyu_tool_test.py [hunk 7]
```diff
@@ -208,6 +259,40 @@ def test_consecutive_spaces(self):
...
+        self.assertIn(["arg1"], iwyu_call_args)
+
+    def test_argument_parser_does_include_xiwyu_flag(self):
+        """ Check that -Xiwyu and --iwyu_opt arguments are passed to the iwyu call. """
+        argv_stub = StubSysArgv(["iwyu_tool.py", "-p", ".", "--", "arg1", "-Xiwyu", "arg2", "--iwyu_opt", "arg3"])
+        iwyu_tool._bootstrap()
+        iwyu_call_args = self.iwyu_tool_main_mock.get_call_args()
+        self.assertIn(["arg1", "-Xiwyu", "arg2", "--iwyu_opt", "arg3"], iwyu_call_args)        
+
+    def test_argument_parser_uses_the_first_separator_for_splitting_arguments(self):
```
Comment (kimgr) [inline iwyu_tool_test.py]: Try shortening this method name so it fits within 80-column width

**Comment by alacasta** [comment]
Hi @kimgr 
Thanks for the review, i will be working on it. 
Before all, i'd like to know if there is a pylint-like, PEP-8 or whatever configuration file with all the requirements about format like this you talk about. It's just to be sure about what style guidelines are required before commiting my code to be reviewed.

File: iwyu_tool_test.py (hunk 2 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️

File: iwyu_tool_test.py (hunk 3 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️

File: iwyu_tool_test.py (hunk 4 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️

File: iwyu_tool_test.py (hunk 5 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️ 
Absolutely right. I misunderstood the meaning of 'iwyu_opt' flag.

File: iwyu_tool_test.py (hunk 7 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️

File: iwyu_tool_test.py (hunk 6 above)
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️ 
# Details:
I like the idea but i think that finding a mechanism for respecting the (probably changing in time) API may allow others not to fall into waste of time trying to find why tests may fail. 
The solution for this has been to inspect the iwyu_tool.main API and from that point, to create two sepparated attributes for the mock, one containing all these arguments and another one (a dictionary) that will be filled with the arguments of the call. 
Hence, you can get the arguments with something like
```python
       call_args = self.iwyu_tool_main_mock.get_call_args()
       self.assertEqual(['arg1'], call_args['iwyu_args'])
```
I liked the possibility of adding the attributes directly to the mock, but i think that accessing the mock attributes with the name of the arguments considering that they are the arguments used in the last call is less standard.
Let's see how do you see this option but i think that is an elegant approach.

File: iwyu_tool.py (hunk 1 above)
Comment (alacasta) [reply iwyu_tool.py]: Done ✔️ 
I like the suggestion. Ii've applied this into the whole iwyu_tool.

File: iwyu_tool.py [hunk 8]
```diff
@@ -435,12 +435,6 @@ def partition_args(argv):
...
     args = parser.parse_args(argv)
-
-    # Force -Xiwyu prefix to iwyu_args so users don't have to provide prefix
-    # explicitly.
-    prefixes = ['-Xiwyu'] * len(iwyu_args)
-    iwyu_args = list(sum(zip(prefixes, iwyu_args), ()))
-
     sys.exit(main(args.dbpath, args.source, args.verbose,
```
[the remark this answers is not in the document]
Comment (alacasta) [reply iwyu_tool.py]: Not applied ❌ (See comment)

File: iwyu_tool_test.py [hunk 9]
```diff
@@ -52,6 +52,57 @@ def will_return(self, content):
...
+        sys.exit = lambda x: self._will_return
+        self._will_return = 0
+
+    def reset(self):
+        sys.exit = self.real_sys_exit
+
+    def will_return(self, content):
+        self._will_return = content   
+
+class StubSysArgv(object):
```
[the remark this answers is not in the document]
Comment (alacasta) [reply iwyu_tool_test.py]: Not applied ❌ (See comment)

**Comment by alacasta** [comment]
Hello @kimgr 
Thanks for the review. I have applied all the changes in three different commits:

1. Typos, style format and unused variables removed
2. Modification on how the iwyu_tool.main() mock register the arguments from a callee has been introduced so that checks are made using directly the API naming instead of 'implicit' order on the call. (further details can be found in the conversation).
3. Refactor of the iwyu_args to extra_args has been applied

However, i consder that the first two changes you propose (bascally, moving sys calls to main) are out of the scope of the current change. From my side, it is not a problem to create this stubs and if in a future is considered better to move sys to main, then we can just remove all this stubs. 
One of my main concerns about this is how the `__main__` is built. There are spare logic in _bootstrap and in main(). IMHO, I'd propose to introduce the _bootstrap logic into main() perhaps splitting all the argument logint in different parts, but at the end, conclding in something like

```python
if __name__ == '__main__':
    sys.exit(main(sys.argv))
``` 
So, I'd like to know your opinion on this:
- Should i definitively remove the sys logic from _bootstrap just not to need to mocked this components as i already did?
- Alternatively, should we keep everything as it and perhaps (if you think that this is needed, i can open an issue for this) this can be handle in a future task

Thanks!

**Comment by kimgr** [comment]
Thanks! I won't have time to look at the patch today because New Year's Eve :fireworks:, but I thought I'd answer some of your questions briefly:

> Before all, i'd like to know if there is a pylint-like, PEP-8 or whatever configuration file with all the
> requirements about format

`iwyu_tool.py` should follow standard PEP8 (with an exception for line-width -- PEP8 defaults to 79, we use 80). `pep8 iwyu_tool.py` finds a few complaints but mostly long lines.

> However, i consder that the first two changes you propose (bascally, moving sys
> calls to main) are out of the scope of the current change. 

Sort of. But they're trivial changes that improve testability, and the diff would be comparatively tiny.

> if __name__ == '__main__':
>    sys.exit(main(sys.argv))

I wouldn't want `main` to take an argv list. As it currently works, you can `import iwyu_tool` as a module and invoke `main` as a regular function. If you change it to take a list, callers would need to compose that list before calling `main`, which would be a weird interface for a function.

> Should i definitively remove the sys logic from _bootstrap just not to need to mocked this components > as i already did?

Yes, I think that would be better. All you need to do is re-shape it as:
```
if __name__ == '__main__':
    sys.exit(_bootstrap(sys.argv))
```

and testability would be much improved. The only reason it wasn't like this already is because `iwyu_tool.py` was originally intended as more of an example/minimal util and didn't have tests. Then it grew unexpectedely.

Happy new year!

**Comment by alacasta** [comment]
Thanks!
Ok, i see the point; i will apply that changes in the implementation now that we have made modifications on this part (tests will keep clearer!).

One additional question:

> `iwyu_tool.py` should follow standard PEP8 (with an exception for line-width -- PEP8 defaults to 79, we use 80). `pep8 iwyu_tool.py` finds a few complaints but mostly long lines.

is then PEP8 used as the official standard for IWYU? If so, should not we set checks on this on CI? I think this would help developers to not ping-ponging due to style related mistakes. 

Happy new year!

File: iwyu_tool_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alacasta) [reply iwyu_tool_test.py]: Done ✔️

File: iwyu_tool.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alacasta) [reply iwyu_tool.py]: Done ✔️

**Comment by alacasta** [comment]
Hello again, 
I've applied the modifications suggested. Basically, all the `sys` dependency has been moved into the `__main__` `_boostrap` call so that it can be easier tested. Moreover, all the stubs created in this PR have been also removed.

**Comment by kimgr** [comment]
I would like a clean commit history before merging this, but I also see several nits in the code itself. Let me try and adopt your patch and shape it to my liking, so we can get this in place. Thanks for the hint on `inspect`, we'll see if it becomes necessary.

**Comment by alacasta** [comment]
Maybe I can try to create one new PR with a couple of commits containing the final implementation. Just let me know.

**Comment by kimgr** [comment]
Thanks, I'd rather build off your excellent work to finalize something without bandying back and forth with review comments. I'll make sure to attribute the work to you.

**Comment by alacasta** [comment]
Looks good! 
The complete `test_argparse_args` use case was missing and is convinient. 

Two minor comments related to style.
- L243 - Missing white-space before 'src1' 
- L262 - Missing white-space before '-Xiwyu'

Thanks!

**Comment by kimgr** [comment]
Thanks, good catch! Please see PR #634 for an updated version. I'll close this in the meantime.
